contracts/Genesis.sol

```diff
@@ -82,107 +91,107 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
-        internal
-        override
-    {
+    function fulfillRandomness(bytes32 requestId, uint256 randomNumber) internal override {
         address nftOwner = requestIdToSender[requestId];
         uint256 tokenId = requestIdToTokenId[requestId];
-        require(
-            _nextTokenId.current() > tokenId,
-            "TokenId has not been minted yet!"
-        );
+        require(_nextTokenId.current() > tokenId, "TokenId has not been minted yet!");
         _mint(nftOwner, tokenId);
-        tokenIdToTokenType[tokenId] = generateRandomTraits(randomNumber);
+        TokenType tokenType = getTokenType(randomNumber);
+        tokenIdToTokenType[tokenId] = tokenType;
+        emit Minted(tokenId, tokenType);
     }
 
     /**
-     * Minting functions
+     * Mint a given number of tokens
+     * The sender has to be included in the free mint list (except if they are the owner)
+     * @param count number of tokens to mint
+     * @param nonce nonce used to verify that the caller is allowed to mint
+     * @param proof Proof to verify that the caller is allowed to mint
      */
-    // TODO Update function to properly use the requests
-    // TODO: Add logic for the reserved spots
-    function mintNFTs(uint256 _count) public payable {
-        require(mintActive, "Minting is not active yet!");
-        uint256 mintIndex = _nextTokenId.current();
-        require(mintIndex + _count <= MAX_SUPPLY, "NFTs sold out");
-        require(msg.value >= PRICE * _count, "Not enough ETH");
-
-        for (uint256 i = 0; i < _count; i++) {
-            // TODO Validate this
-            mintIndex = _nextTokenId.current();
-            _nextTokenId.increment();
-            requestRandomNumberForTokenId(mintIndex);
-        }
+    function freeMint(uint256 count, uint256 nonce, bytes32[] calldata proof) public payable {
+        // TODO
     }
 
-    // TODO: Add logic for the reserved spots
-    function mint() public payable returns (bytes32) {
+    /**
+     * Mint 1 token
+     * The sender has to be included in the whitelist (except if they are the owner)
+     * @param nonce nonce used to verify that the caller is allowed to mint
+     * @param proof Proof to verify that the caller is allowed to mint
+     * @return requestId request id that was sent to Chainlink VRF
+     */
+    function mintWhitelist(uint256 nonce, bytes32[] calldata proof) public payable onlyWhitelist(nonce, proof) returns (bytes32 requestId) {
         require(mintActive, "Minting is not active yet!");
         uint256 mintIndex = _nextTokenId.current();
-        require(mintIndex <= MAX_SUPPLY, "NFTs sold out");
+        require(mintIndex <= MAX_SUPPLY, "Sold out");
         require(msg.value >= PRICE, "Not enough ETH");
-        require(
-            addressToMintCount[msg.sender] < MAX_PER_MINT,
-            "No more minting spot left"
-        );
-
+        require(tokenMintedCount[msg.sender] == 0, "Already minted");
+        tokenMintedCount[msg.sender]++;
```

Comment:
It's safe to say you will never have an overflow here, is that why you didn't use the `SafeMath` method *(`tokenMintedCount[msg.sender] = tokenMintedCount[msg.sender].add(1)`)?

Also wondering if this has a big impact on cost, we can safely assume there will never be an overflow, this way could be more optimal 🤷‍♂️

Reply:
Yep thats why, it will never go more than 1.
I think it's more efficient on gas, but it must be negligible.